cpg_workflows/jobs/sample_batching.py

```diff
@@ -0,0 +1,167 @@
+    n_samples = len(md)
+
+    # shortcut return if the total samples is a valid batch
+    if min_batch_size <= n_samples <= max_batch_size:
+        logging.info(f'Number of samples ({n_samples}) is within range of batch sizes')
+        return [
+            {
+                'samples': md.ID.tolist(),
+                'mf_ratio': is_male / is_female,
+                'size': n_samples,
+                'coverage_medians': md.median_coverage.tolist(),
+            }
+        ]
+
+    # start with a single bin and work upwards
+    cov_bins = 1
+    n_per_batch = n_samples // cov_bins
+
+    # add more bins until we get to a reasonable number of samples per batch
+    # shrink in one direction by increasing num. batches
+    while n_per_batch > max_batch_size:
+        cov_bins += 1
+        n_per_batch = n_samples // cov_bins
+    # endregion
+
+    logging.info(
+        f"""
+    Batching Specs:
+    Total samples: {n_samples}
+    Num. bins: {cov_bins}
+    Approx samples per batch: {n_per_batch}
+"""
+    )
+
+    # Split samples by sex, then roughly by coverage
+    md_sex = {
+        'male': md[~is_female].sort_values(by='median_coverage'),
+        'female': md[is_female].sort_values(by='median_coverage'),
+    }
+
+    # array split each sex proportionally
+    md_sex_cov = {sex: np.array_split(md_sex[sex], cov_bins) for sex in SEX_VALS}
+
+    # create batches
+    batches = []
+    for cov in range(cov_bins):
+        sample_ids = pd.concat([md_sex_cov['male'][cov], md_sex_cov['female'][cov]])
+        batches.append(
+            {
+                'samples': sample_ids.ID.tolist(),
+                'size': len(sample_ids),
+                'mf_ratio': len(md_sex_cov['male'][cov])
+                / len(md_sex_cov['female'][cov]),
+                'coverage_medians': sample_ids.median_coverage.tolist(),
+            }
+        )
+
+    return batches
+
+
+def partition_batches(
+    metadata_file: str,
+    sample_ids: list[str],
+    output_json: str,
+    min_batch_size: int,
+    max_batch_size: int,
+):
+    """
+    Runs this process
+    - loads the input data
+    - obtains batches
+    - writes out the data into one single cross-batch file
+    - also writes out per-batch sample lists
+
+    Args:
+        metadata_file (str): path to the metadata file
+        sample_ids (list[str]): sample IDs to consider
+        output_json (str): location to write the batched samples out
+        min_batch_size (int): minimum batch size
+        max_batch_size (int): maximum batch size
+    """
+
+    # read in the metadata contents
+    logging.basicConfig(level=logging.INFO)
+
+    # load in the metadata file
+    md = pd.read_csv(metadata_file, sep='\t', low_memory=False)
+    md.columns = [x.replace('#', '') for x in md.columns]
+
+    # filter to the PCR-state samples we're interested in
+    # surely there's a neater way to do this...
+    md = md[np.array([sam in sample_ids for sam in md.ID.tolist()])]
```

Comment:
Perhaps `md = md[md['ID'].isin(sample_ids)]` or `md = md.query("ID in @sample_ids")` ?

Reply:
Nice!